File: SuperiorPackGroup/ItemPoolXtraForm.vb
Imports DXDAL.SPGData
Imports DevExpress.Xpo
Imports DevExpress.Data.Filtering
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo
Imports DevExpress.XtraEditors
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.XtraGrid

Public Class ItemPoolXtraForm

    Private m_CurrentPoolID As ItemPool
    Private m_ItemPoolSession As Session

    Private Sub addBarButtonItem_ItemClick(sender As Object, e As DevExpress.XtraBars.ItemClickEventArgs) Handles addBarButtonItem.ItemClick

        m_CurrentPoolID = New ItemPool(m_ItemPoolSession)
        codeTextEdit.Text = Nothing
        customerLookUpEdit.EditValue = Nothing
        customerLookUpEdit.Enabled = True
        descriptionMemoEdit.Text = Nothing

        poolDetailsXpCollection.Criteria = New BinaryOperator(ItemPoolDetails.Fields.ItemPoolID, m_CurrentPoolID.Oid, BinaryOperatorType.Equal)
        poolDetailsXpCollection.Reload()

        If CType(Me.customerLookUpEdit.EditValue, String) = Me.customerLookUpEdit.Properties.NullText Then
            itemRepositoryItemLookUpEdit.DataSource = ItemsBLL.GetItemCodeAndIDsByTypesAndCustomerID(Nothing, "RM", True, m_ItemPoolSession)
        Else
            itemRepositoryItemLookUpEdit.DataSource = ItemsBLL.GetItemCodeAndIDsByTypesAndCustomerID(CType(Me.customerLookUpEdit.EditValue, Integer?), "RM", True, m_ItemPoolSession)
        End If

        Utilities.MakeFormReadOnly(poolXtraTabPage, False)
        poolDetailsGridView.OptionsBehavior.Editable = True
        poolSearchGridControl.Enabled = False
        CheckPermissions()

        cancelBarButtonItem.Enabled = True
        saveBarButtonItem.Enabled = True
        editBarButtonItem.Enabled = False
        addBarButtonItem.Enabled = False
        deleteBarButtonItem.Enabled = False
        deleteGridColumn.Visible = False

    End Sub

    Private Sub cancelBarButtonItem_ItemClick(sender As Object, e As DevExpress.XtraBars.ItemClickEventArgs) Handles cancelBarButtonItem.ItemClick

        CancelChanges()

        CheckPermissions()

        Utilities.MakeFormReadOnly(poolXtraTabPage, True)
        poolDetailsGridView.OptionsBehavior.Editable = False
        poolSearchGridControl.Enabled = True

        cancelBarButtonItem.Enabled = False
        saveBarButtonItem.Enabled = False

    End Sub

    Private Sub customerLookUpEdit_EditValueChanged(sender As Object, e As EventArgs) Handles customerLookUpEdit.EditValueChanged

        If CType(Me.customerLookUpEdit.EditValue, String) = Me.customerLookUpEdit.Properties.NullText Then
            itemRepositoryItemLookUpEdit.DataSource = ItemsBLL.GetItemCodeAndIDsByTypesAndCustomerID(Nothing, "RM", True, m_ItemPoolSession)
        Else
            itemRepositoryItemLookUpEdit.DataSource = ItemsBLL.GetItemCodeAndIDsByTypesAndCustomerID(CType(Me.customerLookUpEdit.EditValue, Integer?), "RM", True, m_ItemPoolSession)
        End If

    End Sub

    Private Sub deleteBarButtonItem_ItemClick(sender As Object, e As DevExpress.XtraBars.ItemClickEventArgs) Handles deleteBarButtonItem.ItemClick

        If MessageBox.Show("Are you sure you want to delete this item pool.", "Delete Item Pool", MessageBoxButtons.YesNo, MessageBoxIcon.Hand) = Windows.Forms.DialogResult.No Then
            Exit Sub
        End If

        If ItemPoolBLL.DeleteItemPool(m_CurrentPoolID.Oid) <> True Then
            MessageBox.Show("The item pool was not deleted.", "Error Encountered", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Exit Sub
        Else
            BindPoolSearch()
        End If

    End Sub

    Private Sub deleteRepositoryItemButtonEdit_Click(sender As Object, e As EventArgs) Handles deleteRepositoryItemButtonEdit.Click

        If MessageBox.Show("Are you sure you want to remove this item.", "Delete Pool Detail", MessageBoxButtons.YesNo, MessageBoxIcon.Hand) = Windows.Forms.DialogResult.No Then
            Exit Sub
        End If

        Dim detail As ItemPoolDetails = m_ItemPoolSession.GetObjectByKey(Of ItemPoolDetails)(poolDetailsGridView.GetFocusedRowCellValue(colOid))
        If Not detail Is Nothing Then
            detail.Delete()
            If detail.IsDeleted = False Then
                MessageBox.Show("The item was not removed.")
            Else
                poolDetailsGridView.DeleteRow(poolDetailsGridView.FocusedRowHandle)
            End If
        Else
            poolDetailsGridView.DeleteRow(poolDetailsGridView.FocusedRowHandle)
        End If

    End Sub

    Private Sub editBarButtonItem_ItemClick(sender As Object, e As DevExpress.XtraBars.ItemClickEventArgs) Handles editBarButtonItem.ItemClick

        If m_CurrentPoolID Is Nothing Then
            MessageBox.Show("Please select a item pool to edit.")
            Exit Sub
        End If

        Utilities.MakeFormReadOnly(poolXtraTabPage, False)
        poolDetailsGridView.OptionsBehavior.Editable = True
        poolSearchGridControl.Enabled = False
        CheckPermissions()

        cancelBarButtonItem.Enabled = True
        saveBarButtonItem.Enabled = True
        editBarButtonItem.Enabled = False
        addBarButtonItem.Enabled = False
        deleteBarButtonItem.Enabled = False

    End Sub

    Private Sub ItemPoolXtraForm_Activated(sender As Object, e As EventArgs) Handles Me.Activated

        XPBaseObject.AutoSaveOnEndEdit = False

    End Sub

    Private Sub ItemPoolXtraForm_FormClosing(sender As Object, e As FormClosingEventArgs) Handles MyBase.FormClosing

        If poolSearchGridControl.Enabled = False Then
            Select Case MessageBox.Show("Do you want to save changes?", "Save Changes", MessageBoxButtons.YesNoCancel, MessageBoxIcon.Question)
                Case DialogResult.Yes
                    If SaveChanges() = False Then
                        e.Cancel = True
                    End If
                Case DialogResult.No
                    CancelChanges()
                Case DialogResult.Cancel
                    e.Cancel = True
            End Select
        End If

        If e.Cancel = False Then
            XPBaseObject.AutoSaveOnEndEdit = True
            m_ItemPoolSession.PurgeDeletedObjects()
        End If

    End Sub

    Private Sub ItemPoolXtraForm_Load(sender As Object, e As EventArgs) Handles Me.Load

        Cursor = Cursors.WaitCursor

        CheckPermissions()
        m_ItemPoolSession = New Session(SPGDataLayer)
        m_ItemPoolSession.TrackPropertiesModifications = True
        m_ItemPoolSession.OptimisticLockingReadBehavior = OptimisticLockingReadBehavior.MergeCollisionThrowException
        poolDetailsXpCollection.Session = m_ItemPoolSession
        poolSearchXpView.Session = m_ItemPoolSession
        itemXpView.Session = m_ItemPoolSession
        customerXpView.Session = m_ItemPoolSession

        BindPoolSearch()

        poolDetailsXpCollection.Criteria = New BinaryOperator(ItemPoolDetails.Fields.ItemPoolID, 0, BinaryOperatorType.Equal)

        Utilities.MakeFormReadOnly(poolXtraTabPage, True)
        poolDetailsGridView.OptionsBehavior.Editable = False
        poolSearchGridControl.Enabled = True

        cancelBarButtonItem.Enabled = False
        saveBarButtonItem.Enabled = False

        Cursor = Cursors.Default

    End Sub

    Private Sub itemRepositoryItemLookUpEdit_Closed(sender As Object, e As DevExpress.XtraEditors.Controls.ClosedEventArgs) Handles itemRepositoryItemLookUpEdit.Closed

        If e.CloseMode <> PopupCloseMode.Cancel Then
            Dim view As GridView = CType(CType(CType(sender, LookUpEdit).Parent, GridControl).FocusedView, GridView)
            If view.IsNewItemRow(view.FocusedRowHandle) Then
                view.CloseEditor()
                view.UpdateCurrentRow()
            End If
        End If

    End Sub

    Private Sub poolDetailsGridView_CustomUnboundColumnData(sender As Object, e As Views.Base.CustomColumnDataEventArgs) Handles poolDetailsGridView.CustomUnboundColumnData

        If Not IsDBNull(poolDetailsGridView.GetListSourceRowCellValue(e.ListSourceRowIndex, itemGridColumn)) AndAlso poolDetailsGridView.GetListSourceRowCellValue(e.ListSourceRowIndex, itemGridColumn) IsNot Nothing _
               AndAlso Not IsDBNull(poolDetailsGridView.GetListSourceRowCellValue(e.ListSourceRowIndex, itemQtyOnHandGridColumn)) Then
            If e.Column.Name = poolQuantityOnHandGridColumn.Name And e.IsGetData Then
                e.Value = Convert.ToDouble(poolDetailsGridView.GetListSourceRowCellValue(e.ListSourceRowIndex, itemQtyOnHandGridColumn)) * (Convert.ToDouble(poolDetailsGridView.GetListSourceRowCellValue(e.ListSourceRowIndex, colItemPoolRatio)))
            End If
        End If

    End Sub

    Private Sub poolDetailsGridView_InitNewRow(sender As Object, e As InitNewRowEventArgs) Handles poolDetailsGridView.InitNewRow

        poolDetailsGridView.SetRowCellValue(e.RowHandle, colPoolID, m_CurrentPoolID.Oid)
        poolDetailsGridView.SetRowCellValue(e.RowHandle, colItemPoolRatio, 1)

    End Sub

    Private Sub poolSearchGridView_Click(sender As Object, e As EventArgs) Handles poolSearchGridView.Click

        If poolSearchGridView.FocusedRowHandle = -999997 Then
            Exit Sub
        End If

        If (poolSearchGridView.CalcHitInfo(poolSearchGridControl.PointToClient(Control.MousePosition)).HitTest = GridHitTest.RowCell) Then
            BindPoolControls(CType(poolSearchGridView.GetFocusedRowCellValue(colPoolID), Integer))
        End If

    End Sub

    Private Sub saveBarButtonItem_ItemClick(sender As Object, e As DevExpress.XtraBars.ItemClickEventArgs) Handles saveBarButtonItem.ItemClick

        If SaveChanges() Then
            CheckPermissions()

            Utilities.MakeFormReadOnly(poolXtraTabPage, True)
            poolDetailsGridView.OptionsBehavior.Editable = False
            poolSearchGridControl.Enabled = True

            cancelBarButtonItem.Enabled = False
            saveBarButtonItem.Enabled = False
        End If

    End Sub

    Public Sub BindPoolControls(ByVal poolID As Integer)

        m_ItemPoolSession.DropIdentityMap()

        Dim pool As ItemPool = m_ItemPoolSession.GetObjectByKey(Of ItemPool)(poolID, True)
        m_CurrentPoolID = pool
        codeTextEdit.Text = pool.PoolCode
        customerLookUpEdit.EditValue = pool.PoolCustomerID.CustomerID
        customerLookUpEdit.Enabled = False
        descriptionMemoEdit.Text = pool.PoolDescription

        poolDetailsXpCollection.Criteria = New BinaryOperator(ItemPoolDetails.Fields.ItemPoolID, m_CurrentPoolID.Oid, BinaryOperatorType.Equal)

        poolDetailsXpCollection.Reload()

        If CType(Me.customerLookUpEdit.EditValue, String) = Me.customerLookUpEdit.Properties.NullText Then
            itemRepositoryItemLookUpEdit.DataSource = ItemsBLL.GetItemCodeAndIDsByTypesAndCustomerID(Nothing, "RM", True, m_ItemPoolSession)
        Else
            itemRepositoryItemLookUpEdit.DataSource = ItemsBLL.GetItemCodeAndIDsByTypesAndCustomerID(CType(customerLookUpEdit.EditValue, Integer?), "RM", True, m_ItemPoolSession)
        End If

    End Sub

    Public Sub BindPoolSearch()

        poolSearchXpView.Reload()

    End Sub

    Private Sub CancelChanges()

        If m_CurrentPoolID.Oid <> -1 Then
            BindPoolControls(m_CurrentPoolID.Oid)
        End If

    End Sub

    Public Function CheckPermissions() As Boolean

        Try
            Dim permission As SPG.UserPermissionsRow = New UserPermissionsBLL().GetUserPermissionsByUserName(My.Settings.UserName, "Items Form")(0)

            Select Case permission.PermissionLevel
                Case "FULL", "Full"
                    Me.addBarButtonItem.Enabled = True
                    Me.editBarButtonItem.Enabled = True
                    Me.deleteBarButtonItem.Enabled = True
                    Me.deleteGridColumn.Visible = True
                    Me.deleteGridColumn.VisibleIndex = 0
                Case "ADD", "Add"
                    Me.addBarButtonItem.Enabled = True
                    Me.editBarButtonItem.Enabled = True
                    Me.deleteBarButtonItem.Enabled = False
                    Me.deleteGridColumn.Visible = False
                Case "NONE", "None"
                    MessageBox.Show("You don't have permission to view this form.", "User Permissions", MessageBoxButtons.OK, MessageBoxIcon.Hand)
                    Me.Close()
                    Return False
                Case "VIEW", "View"
                    Me.addBarButtonItem.Enabled = False
                    Me.editBarButtonItem.Enabled = False
                    Me.deleteBarButtonItem.Enabled = False
                    Me.deleteGridColumn.Visible = False
                Case "EDIT", "Edit"
                    Me.addBarButtonItem.Enabled = False
                    Me.editBarButtonItem.Enabled = True
                    Me.deleteBarButtonItem.Enabled = False
                    Me.deleteGridColumn.Visible = False
                Case "VIEW ASSIGNED"
                    Me.addBarButtonItem.Enabled = False
                    Me.editBarButtonItem.Enabled = False
                    Me.deleteGridColumn.Visible = False
                    Me.deleteBarButtonItem.Enabled = False
                    FilterAssignedCustomers()
                Case "EDIT ASSIGNED"
                    Me.addBarButtonItem.Enabled = False
                    Me.editBarButtonItem.Enabled = True
                    Me.deleteGridColumn.Visible = False
                    Me.deleteBarButtonItem.Enabled = False
                    FilterAssignedCustomers()
                Case Else
                    MessageBox.Show("You don't have permission to view this form.", "User Permissions", MessageBoxButtons.OK, MessageBoxIcon.Error)
                    Me.Close()
                    Return False
            End Select
            Return True
        Catch ex As InvalidOperationException
            Return False
        End Try

    End Function

    Private Sub FilterAssignedCustomers()

        Me.poolSearchXpView.Criteria = New InOperator(ItemPool.Fields.PoolCustomerID.CustomerID.PropertyName, UsersCustomerBLL.GetAssignedCustomers(m_ItemPoolSession))

    End Sub

    Private Function SaveChanges() As Boolean

        Dim selectedID As Integer = -1
        If m_CurrentPoolID.Oid = -1 Then

        Else
            selectedID = poolSearchGridView.FocusedRowHandle
        End If

        If ItemPoolBLL.ValidateRecord(Me.codeTextEdit.Text, CType(Me.customerLookUpEdit.EditValue, Integer?)) Then
            m_CurrentPoolID.PoolCode = codeTextEdit.Text
            m_CurrentPoolID.PoolDescription = descriptionMemoEdit.Text
            m_CurrentPoolID.PoolCustomerID = CustomersBLL.GetCustomer(CType(customerLookUpEdit.EditValue, Integer), m_ItemPoolSession)
            m_CurrentPoolID.PoolType = "BP"
        Else
            MessageBox.Show("You must provide a Pool Code, and Customer name before saving the record.")
            Return False
        End If

        Try
            m_CurrentPoolID.Save()
        Catch ex As ApplicationException
            MessageBox.Show(ex.Message, "Error Encountered", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return False
        End Try

        'Save the details here
        For Each poolDetail As ItemPoolDetails In poolDetailsXpCollection
            If poolDetail.ItemPoolID Is Nothing OrElse poolDetail.ItemPoolID.Oid = -1 Then
                poolDetail.ItemPoolID = m_CurrentPoolID
            End If
            poolDetail.Save()
        Next

        BindPoolSearch()
        BindPoolControls(m_CurrentPoolID.Oid)
        poolSearchGridView.FocusedRowHandle = selectedID
        Return True

    End Function

End Class

File: SuperiorPackGroup/ItemPoolXtraForm.cs
//INSTANT C# NOTE: Formerly VB project-level imports:
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Drawing;
using System.Diagnostics;
using System.Windows.Forms;

using DXDAL.SPGData;
using DevExpress.Xpo;
using DevExpress.Data.Filtering;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;
using DevExpress.XtraEditors;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.XtraGrid;

namespace SuperiorPackGroup
{
	public partial class ItemPoolXtraForm
	{

		public ItemPoolXtraForm()
		{
			InitializeComponent();
		}

		private ItemPool m_CurrentPoolID;
		private Session m_ItemPoolSession;

		private void addBarButtonItem_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e)
		{

			m_CurrentPoolID = new ItemPool(m_ItemPoolSession);
			codeTextEdit.Text = null;
			customerLookUpEdit.EditValue = null;
			customerLookUpEdit.Enabled = true;
			descriptionMemoEdit.Text = null;

			poolDetailsXpCollection.Criteria = new BinaryOperator(ItemPoolDetails.Fields.ItemPoolID, m_CurrentPoolID.Oid, BinaryOperatorType.Equal);
			poolDetailsXpCollection.Reload();

			if ((customerLookUpEdit.EditValue == null ? null : Convert.ToString(customerLookUpEdit.EditValue)) == customerLookUpEdit.Properties.NullText)
			{
				itemRepositoryItemLookUpEdit.DataSource = ItemsBLL.GetItemCodeAndIDsByTypesAndCustomerID(null, "RM", true, m_ItemPoolSession);
			}
			else
			{
				itemRepositoryItemLookUpEdit.DataSource = ItemsBLL.GetItemCodeAndIDsByTypesAndCustomerID((int?)customerLookUpEdit.EditValue, "RM", true, m_ItemPoolSession);
			}

			Utilities.MakeFormReadOnly(poolXtraTabPage, false);
			poolDetailsGridView.OptionsBehavior.Editable = true;
			poolSearchGridControl.Enabled = false;
			CheckPermissions();

			cancelBarButtonItem.Enabled = true;
			saveBarButtonItem.Enabled = true;
			editBarButtonItem.Enabled = false;
			addBarButtonItem.Enabled = false;
			deleteBarButtonItem.Enabled = false;
			deleteGridColumn.Visible = false;

		}

		private void cancelBarButtonItem_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e)
		{

			CancelChanges();

			CheckPermissions();

			Utilities.MakeFormReadOnly(poolXtraTabPage, true);
			poolDetailsGridView.OptionsBehavior.Editable = false;
			poolSearchGridControl.Enabled = true;

			cancelBarButtonItem.Enabled = false;
			saveBarButtonItem.Enabled = false;

		}

		private void customerLookUpEdit_EditValueChanged(object sender, EventArgs e)
		{

			if ((customerLookUpEdit.EditValue == null ? null : Convert.ToString(customerLookUpEdit.EditValue)) == customerLookUpEdit.Properties.NullText)
			{
				itemRepositoryItemLookUpEdit.DataSource = ItemsBLL.GetItemCodeAndIDsByTypesAndCustomerID(null, "RM", true, m_ItemPoolSession);
			}
			else
			{
				itemRepositoryItemLookUpEdit.DataSource = ItemsBLL.GetItemCodeAndIDsByTypesAndCustomerID((int?)customerLookUpEdit.EditValue, "RM", true, m_ItemPoolSession);
			}

		}

		private void deleteBarButtonItem_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e)
		{

			if (MessageBox.Show("Are you sure you want to delete this item pool.", "Delete Item Pool", MessageBoxButtons.YesNo, MessageBoxIcon.Hand) == DialogResult.No)
			{
				return;
			}

			if (ItemPoolBLL.DeleteItemPool(m_CurrentPoolID.Oid) != true)
			{
				MessageBox.Show("The item pool was not deleted.", "Error Encountered", MessageBoxButtons.OK, MessageBoxIcon.Error);
				return;
			}
			else
			{
				BindPoolSearch();
			}

		}

		private void deleteRepositoryItemButtonEdit_Click(object sender, EventArgs e)
		{

			if (MessageBox.Show("Are you sure you want to remove this item.", "Delete Pool Detail", MessageBoxButtons.YesNo, MessageBoxIcon.Hand) == DialogResult.No)
			{
				return;
			}

			ItemPoolDetails detail = m_ItemPoolSession.GetObjectByKey<ItemPoolDetails>(poolDetailsGridView.GetFocusedRowCellValue(colOid));
			if (detail != null)
			{
				detail.Delete();
				if (detail.IsDeleted == false)
				{
					MessageBox.Show("The item was not removed.");
				}
				else
				{
					poolDetailsGridView.DeleteRow(poolDetailsGridView.FocusedRowHandle);
				}
			}
			else
			{
				poolDetailsGridView.DeleteRow(poolDetailsGridView.FocusedRowHandle);
			}

		}

		private void editBarButtonItem_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e)
		{

			if (m_CurrentPoolID == null)
			{
				MessageBox.Show("Please select a item pool to edit.");
				return;
			}

			Utilities.MakeFormReadOnly(poolXtraTabPage, false);
			poolDetailsGridView.OptionsBehavior.Editable = true;
			poolSearchGridControl.Enabled = false;
			CheckPermissions();

			cancelBarButtonItem.Enabled = true;
			saveBarButtonItem.Enabled = true;
			editBarButtonItem.Enabled = false;
			addBarButtonItem.Enabled = false;
			deleteBarButtonItem.Enabled = false;

		}

		private void ItemPoolXtraForm_Activated(object sender, EventArgs e)
		{

			XPBaseObject.AutoSaveOnEndEdit = false;

		}

		private void ItemPoolXtraForm_FormClosing(object sender, FormClosingEventArgs e)
		{

			if (poolSearchGridControl.Enabled == false)
			{
				switch (MessageBox.Show("Do you want to save changes?", "Save Changes", MessageBoxButtons.YesNoCancel, MessageBoxIcon.Question))
				{
					case DialogResult.Yes:
						if (SaveChanges() == false)
						{
							e.Cancel = true;
						}
						break;
					case DialogResult.No:
						CancelChanges();
						break;
					case DialogResult.Cancel:
						e.Cancel = true;
						break;
				}
			}

			if (e.Cancel == false)
			{
				XPBaseObject.AutoSaveOnEndEdit = true;
				m_ItemPoolSession.PurgeDeletedObjects();
			}

		}

		private void ItemPoolXtraForm_Load(object sender, EventArgs e)
		{

			Cursor = Cursors.WaitCursor;

			CheckPermissions();
			m_ItemPoolSession = new Session(MyDataLayers.SPGDataLayer);
			m_ItemPoolSession.TrackPropertiesModifications = true;
			m_ItemPoolSession.OptimisticLockingReadBehavior = OptimisticLockingReadBehavior.MergeCollisionThrowException;
			poolDetailsXpCollection.Session = m_ItemPoolSession;
			poolSearchXpView.Session = m_ItemPoolSession;
			itemXpView.Session = m_ItemPoolSession;
			customerXpView.Session = m_ItemPoolSession;

			BindPoolSearch();

			poolDetailsXpCollection.Criteria = new BinaryOperator(ItemPoolDetails.Fields.ItemPoolID, 0, BinaryOperatorType.Equal);

			Utilities.MakeFormReadOnly(poolXtraTabPage, true);
			poolDetailsGridView.OptionsBehavior.Editable = false;
			poolSearchGridControl.Enabled = true;

			cancelBarButtonItem.Enabled = false;
			saveBarButtonItem.Enabled = false;

			Cursor = Cursors.Default;

		}

		private void itemRepositoryItemLookUpEdit_Closed(object sender, DevExpress.XtraEditors.Controls.ClosedEventArgs e)
		{

			if (e.CloseMode != PopupCloseMode.Cancel)
			{
				GridView view = (GridView)(((GridControl)(((LookUpEdit)sender).Parent)).FocusedView);
				if (view.IsNewItemRow(view.FocusedRowHandle))
				{
					view.CloseEditor();
					view.UpdateCurrentRow();
				}
			}

		}

		private void poolDetailsGridView_CustomUnboundColumnData(object sender, DevExpress.XtraGrid.Views.Base.CustomColumnDataEventArgs e)
		{

			if (!Convert.IsDBNull(poolDetailsGridView.GetListSourceRowCellValue(e.ListSourceRowIndex, itemGridColumn)) && poolDetailsGridView.GetListSourceRowCellValue(e.ListSourceRowIndex, itemGridColumn) != null && !Convert.IsDBNull(poolDetailsGridView.GetListSourceRowCellValue(e.ListSourceRowIndex, itemQtyOnHandGridColumn)))
			{
				if (e.Column.Name == poolQuantityOnHandGridColumn.Name && e.IsGetData)
				{
					e.Value = Convert.ToDouble(poolDetailsGridView.GetListSourceRowCellValue(e.ListSourceRowIndex, itemQtyOnHandGridColumn)) * (Convert.ToDouble(poolDetailsGridView.GetListSourceRowCellValue(e.ListSourceRowIndex, colItemPoolRatio)));
				}
			}

		}

		private void poolDetailsGridView_InitNewRow(object sender, InitNewRowEventArgs e)
		{

			poolDetailsGridView.SetRowCellValue(e.RowHandle, colPoolID, m_CurrentPoolID.Oid);
			poolDetailsGridView.SetRowCellValue(e.RowHandle, colItemPoolRatio, 1);

		}

		private void poolSearchGridView_Click(object sender, EventArgs e)
		{

			if (poolSearchGridView.FocusedRowHandle == -999997)
			{
				return;
			}

			if (poolSearchGridView.CalcHitInfo(poolSearchGridControl.PointToClient(Control.MousePosition)).HitTest == GridHitTest.RowCell)
			{
				BindPoolControls(Convert.ToInt32(poolSearchGridView.GetFocusedRowCellValue(colPoolID)));
			}

		}

		private void saveBarButtonItem_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e)
		{

			if (SaveChanges())
			{
				CheckPermissions();

				Utilities.MakeFormReadOnly(poolXtraTabPage, true);
				poolDetailsGridView.OptionsBehavior.Editable = false;
				poolSearchGridControl.Enabled = true;

				cancelBarButtonItem.Enabled = false;
				saveBarButtonItem.Enabled = false;
			}

		}

		public void BindPoolControls(int poolID)
		{

			m_ItemPoolSession.DropIdentityMap();

			ItemPool pool = m_ItemPoolSession.GetObjectByKey<ItemPool>(poolID, true);
			m_CurrentPoolID = pool;
			codeTextEdit.Text = pool.PoolCode;
			customerLookUpEdit.EditValue = pool.PoolCustomerID.CustomerID;
			customerLookUpEdit.Enabled = false;
			descriptionMemoEdit.Text = pool.PoolDescription;

			poolDetailsXpCollection.Criteria = new BinaryOperator(ItemPoolDetails.Fields.ItemPoolID, m_CurrentPoolID.Oid, BinaryOperatorType.Equal);

			poolDetailsXpCollection.Reload();

			if ((customerLookUpEdit.EditValue == null ? null : Convert.ToString(customerLookUpEdit.EditValue)) == customerLookUpEdit.Properties.NullText)
			{
				itemRepositoryItemLookUpEdit.DataSource = ItemsBLL.GetItemCodeAndIDsByTypesAndCustomerID(null, "RM", true, m_ItemPoolSession);
			}
			else
			{
				itemRepositoryItemLookUpEdit.DataSource = ItemsBLL.GetItemCodeAndIDsByTypesAndCustomerID((int?)customerLookUpEdit.EditValue, "RM", true, m_ItemPoolSession);
			}

		}

		public void BindPoolSearch()
		{

			poolSearchXpView.Reload();

		}

		private void CancelChanges()
		{

			if (m_CurrentPoolID.Oid != -1)
			{
				BindPoolControls(m_CurrentPoolID.Oid);
			}

		}

		public bool CheckPermissions()
		{

			try
			{
				SPG.UserPermissionsRow permission = (new UserPermissionsBLL()).GetUserPermissionsByUserName(Properties.Settings.Default.UserName, "Items Form")[0];

				switch (permission.PermissionLevel)
				{
					case "FULL":
					case "Full":
						addBarButtonItem.Enabled = true;
						editBarButtonItem.Enabled = true;
						deleteBarButtonItem.Enabled = true;
						deleteGridColumn.Visible = true;
						deleteGridColumn.VisibleIndex = 0;
						break;
					case "ADD":
					case "Add":
						addBarButtonItem.Enabled = true;
						editBarButtonItem.Enabled = true;
						deleteBarButtonItem.Enabled = false;
						deleteGridColumn.Visible = false;
						break;
					case "NONE":
					case "None":
						MessageBox.Show("You don't have permission to view this form.", "User Permissions", MessageBoxButtons.OK, MessageBoxIcon.Hand);
						Close();
						return false;
					case "VIEW":
					case "View":
						addBarButtonItem.Enabled = false;
						editBarButtonItem.Enabled = false;
						deleteBarButtonItem.Enabled = false;
						deleteGridColumn.Visible = false;
						break;
					case "EDIT":
					case "Edit":
						addBarButtonItem.Enabled = false;
						editBarButtonItem.Enabled = true;
						deleteBarButtonItem.Enabled = false;
						deleteGridColumn.Visible = false;
						break;
					case "VIEW ASSIGNED":
						addBarButtonItem.Enabled = false;
						editBarButtonItem.Enabled = false;
						deleteGridColumn.Visible = false;
						deleteBarButtonItem.Enabled = false;
						FilterAssignedCustomers();
						break;
					case "EDIT ASSIGNED":
						addBarButtonItem.Enabled = false;
						editBarButtonItem.Enabled = true;
						deleteGridColumn.Visible = false;
						deleteBarButtonItem.Enabled = false;
						FilterAssignedCustomers();
						break;
					default:
						MessageBox.Show("You don't have permission to view this form.", "User Permissions", MessageBoxButtons.OK, MessageBoxIcon.Error);
						Close();
						return false;
				}
				return true;
			}
			catch (InvalidOperationException ex)
			{
				return false;
			}

		}

		private void FilterAssignedCustomers()
		{

			poolSearchXpView.Criteria = new InOperator(ItemPool.Fields.PoolCustomerID.CustomerID.PropertyName, UsersCustomerBLL.GetAssignedCustomers(m_ItemPoolSession));

		}

		private bool SaveChanges()
		{

			int selectedID = -1;
			if (m_CurrentPoolID.Oid == -1)
			{

			}
			else
			{
				selectedID = poolSearchGridView.FocusedRowHandle;
			}

			if (ItemPoolBLL.ValidateRecord(codeTextEdit.Text, (int?)customerLookUpEdit.EditValue))
			{
				m_CurrentPoolID.PoolCode = codeTextEdit.Text;
				m_CurrentPoolID.PoolDescription = descriptionMemoEdit.Text;
				m_CurrentPoolID.PoolCustomerID = CustomersBLL.GetCustomer(Convert.ToInt32(customerLookUpEdit.EditValue), m_ItemPoolSession);
				m_CurrentPoolID.PoolType = "BP";
			}
			else
			{
				MessageBox.Show("You must provide a Pool Code, and Customer name before saving the record.");
				return false;
			}

			try
			{
				m_CurrentPoolID.Save();
			}
			catch (ApplicationException ex)
			{
				MessageBox.Show(ex.Message, "Error Encountered", MessageBoxButtons.OK, MessageBoxIcon.Error);
				return false;
			}

			//Save the details here
			foreach (ItemPoolDetails poolDetail in poolDetailsXpCollection)
			{
				if (poolDetail.ItemPoolID == null || poolDetail.ItemPoolID.Oid == -1)
				{
					poolDetail.ItemPoolID = m_CurrentPoolID;
				}
				poolDetail.Save();
			}

			BindPoolSearch();
			BindPoolControls(m_CurrentPoolID.Oid);
			poolSearchGridView.FocusedRowHandle = selectedID;
			return true;

		}

	}
}